### PR TITLE
Feature #251: RSS feed in JSON format

### PR DESCRIPTION
**Pull request by kraluk**
My proposal of presenting RSS feed in the JSON format.

And want to only notice that I'm not sure if I did not too complicate the `BlogPostsControllerSpec` test - please don't shout at me :)

### CODE REVIEW

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import static com.jvm_bloggers.utils.NowProvider.DEFAULT_ZONE;
+
+/**
+ * Simple {@link SyndFeed} objects to JSON plain text converter
+ *
+ * @author kraluk
+ */
+@Component
+@Slf4j
+public class RssToJsonConverter {
```
Comment (tdziurko) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: If you name it `SyndFeedToJsonConverter` no javadoc would be necessary :)

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    private static JSONObject toJson(SyndEntry entry) {
+        JSONObject json = new JSONObject();
+
+        json.put(LINK.getKey(), entry.getLink());
+        json.put(TITLE.getKey(), entry.getTitle());
+        json.put(AUTHOR.getKey(), entry.getAuthor());
+
+        // Description
+        Option.of(entry.getDescription())
+            .peek(d -> json.put(DESCRIPTION.getKey(), d.getValue()));
```
Comment (tdziurko) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: T.I.L. (Today I Learned) that `peek` is also in `Option` class :) :+1:

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+        // Date
```
Comment (tdziurko) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: Does this comment tells reader more than `String date` below? :)

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        String date = DATE_FORMATTER.format(
```
Comment (tdziurko) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: Maybe you can use `com.jvm_bloggers.utils.DateTimeUtilities` class or add this logic there?

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            LocalDateTime.ofInstant(entry.getPublishedDate().toInstant(), DEFAULT_ZONE));
+
+        json.put(DATE.getKey(), date);
+
```
Comment (tdziurko) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: TooManyEmptyLinexException ;)

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        return json;
+    }
+
+    /**
+     * Converts complete {@link SyndFeed} instance to a JSON form
+     *
+     * @param feed a {@link SyndFeed} instance
+     * @return a {@link JSONObject} containing complete JSONfied feed
+     */
+    public JSONObject convert(SyndFeed feed) {
```
Comment (tdziurko) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: Again, reading this javadoc didn't teach me anything more than I could read from method signature :)

File: src/test/groovy/com/jvm_bloggers/core/rss/BlogPostsControllerSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+    RssToJsonConverter rssToJsonConverter = new RssToJsonConverter("http://jvm-bloggers.com")
+
+    @Subject
+    BlogPostsController blogPostsController = new BlogPostsController(rssProducerMock, rssToJsonConverter, 50)
+
+    @Unroll
+    def "Should get OK status for RSS feed in #format format request"() {
+        given:
+            MockMvc mockMvc = standaloneSetup(blogPostsController).build()
```
Comment (tdziurko) [inline src/test/groovy/com/jvm_bloggers/core/rss/BlogPostsControllerSpec.groovy]: No indentation after labels, convention was simplified to

    given:
    something;
    
    when:
    something;

    then:
    something;

Docs should be updated, but if you have found it somewhere, blame is on me :)

File: src/test/groovy/com/jvm_bloggers/core/rss/BlogPostsControllerSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+        expect:
+            mockMvc.perform(get("/pl/rss?format=$format"))
+                    .andExpect(status().isOk())
+                    .andExpect(content().contentType(mediaType))
+
+        where:
+            format || mediaType
+            "json" || MediaType.APPLICATION_JSON_VALUE
```
Comment (tdziurko) [inline src/test/groovy/com/jvm_bloggers/core/rss/BlogPostsControllerSpec.groovy]: Static imports :)

File: src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java [hunk 9]
```diff
@@ -19,23 +23,53 @@
...
+                       PrintWriter writer,
+                       @RequestParam(required = false) Integer limit,
+                       @RequestParam(required = false) Set<String> excludedAuthors,
+                       @RequestParam(required = false, defaultValue = "xml") String format) {
+        Integer checkedLimit = firstNonNull(limit, defaultLimit);
+
+        SyndFeed feed =
+            rssProducer.getRss(request.getRequestURL().toString(), checkedLimit, excludedAuthors);
+
+        if (SupportedFormat.XML.toString().equalsIgnoreCase(format)) {
```
Comment (airborn) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: I'd also consider splitting this into to separate methods each with it's own `@RequestMapping` like:
`@RequestMapping(path=RSS_FEED_MAPPING, produces = "application/json")`

File: src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java [hunk 10]
```diff
@@ -19,23 +23,53 @@
...
+            response.setContentType(MediaType.APPLICATION_ATOM_XML_VALUE);
+
+            new SyndFeedOutput().output(feed, writer);
+        } else if (SupportedFormat.JSON.toString().equalsIgnoreCase(format)) {
+            response.setContentType(MediaType.APPLICATION_JSON_VALUE);
+
+            rssToJsonConverter.convert(feed).write(writer);
+        } else {
+            response.sendError(HttpStatus.BAD_REQUEST.value(), "Unsupported format type!");
```
Comment (airborn) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: 2. Not sure what if there is a Spring support for this but such a call should be handle by OPTIONS call

**Comment by kraluk** [comment]
@tdziurko @airborn big thanks for the review and comments - I'll try to fix whole issues as soon as possible. Sorry for late response.

**Comment by tdziurko** [comment]
No problem, but treat those as suggestions not orders to follow blindly :) I am open for a discussion if you think that some comments are incorrect.

**Comment by tdziurko** [comment]
Any update on this? :)

**Comment by kraluk** [comment]
Argh... Completely forget about it... Mea culpa!
This weekend I'll pull proper fixes.

File: src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java (hunk 9 above)
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: Finally I decided to play with Vavr this time.

File: src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java [hunk 11]
```diff
@@ -19,23 +23,53 @@
...
+        SyndFeed feed =
+            rssProducer.getRss(request.getRequestURL().toString(), checkedLimit, excludedAuthors);
+
+        if (SupportedFormat.XML.toString().equalsIgnoreCase(format)) {
+            response.setContentType(MediaType.APPLICATION_ATOM_XML_VALUE);
+
+            new SyndFeedOutput().output(feed, writer);
+        } else if (SupportedFormat.JSON.toString().equalsIgnoreCase(format)) {
+            response.setContentType(MediaType.APPLICATION_JSON_VALUE);
+
```
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: Removed :)

File: src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java (hunk 10 above)
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: Done.

Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: OPTIONS call, as I know (but there is a chance that I'm wrong), is not supported by default by Spring, but maybe it'll be a good idea to globally apply such a support to the application as a separate issue?

File: src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java [hunk 12]
```diff
@@ -19,23 +23,53 @@
...
+            rssToJsonConverter.convert(feed).write(writer);
+        } else {
+            response.sendError(HttpStatus.BAD_REQUEST.value(), "Unsupported format type!");
+        }
     }
 
+    /**
+     * Supported RSS feed formats
+     */
+    enum SupportedFormat {
```
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: Done.

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java (hunk 1 above)
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: Good idea, fixed.

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    private static JSONObject toJson(SyndFeed feed) {
+        JSONObject json = new JSONObject();
+
+        json.put(TITLE.getKey(), feed.getTitle());
+        json.put(LINK.getKey(), feed.getLink());
+        json.put(DESCRIPTION.getKey(), feed.getDescription());
+        json.put(ENTRIES.getKey(), toJson(feed.getEntries()));
+
```
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: Removed :)

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        return json;
+    }
+
+    private static java.util.List<JSONObject> toJson(java.util.List<SyndEntry> entries) {
+        return javaslang.collection.List.ofAll(entries)
```
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: Yep, frankly speaking, I'm not sure now, what I've wanted to do here... It was over-engineered completely. At the end, staying with java.util.List (as external libraries required it) I hope it'll be a good idea here, as you suggested :)

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java (hunk 2 above)
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: :)

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java (hunk 3 above)
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: You're completely right - totally unnecessary comment...

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java (hunk 4 above)
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: Thanks for the hint. Fixed.

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java (hunk 5 above)
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: I hope that I've removed all those lines...

File: src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java (hunk 6 above)
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssToJsonConverter.java]: Removed :)

File: src/main/java/com/jvm_bloggers/core/rss/json/RssJsonKey.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
+package com.jvm_bloggers.core.rss.json;
+
+/**
+ * Keys which are used in parsing RSS Feed to a JSON text content
+ *
+ * @author kraluk
+ */
+enum RssJsonKey {
```
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/json/RssJsonKey.java]: You're right - enum here is a symptom of some kind of over-engineering...

File: src/test/groovy/com/jvm_bloggers/core/rss/BlogPostsControllerSpec.groovy (hunk 7 above)
Comment (kraluk) [inline src/test/groovy/com/jvm_bloggers/core/rss/BlogPostsControllerSpec.groovy]: Fixed :)

File: src/test/groovy/com/jvm_bloggers/core/rss/BlogPostsControllerSpec.groovy (hunk 8 above)
Comment (kraluk) [inline src/test/groovy/com/jvm_bloggers/core/rss/BlogPostsControllerSpec.groovy]: Fixed :)

**Comment by kraluk** [comment]
Finally pushed fixes...

But there are some issues discovered by Codacy in the RssJsonKey class (it contains only some constants) - maybe moving those constants directly to the SyndFeedToJsonConverter will be a better idea instead of storing those keys in a separate class?

**Comment by airborn** [comment]
You're trying to merge dozens of commits that are already merged into master. Please rebase on latest master and ensure that only Yours commits are on the branch.

**Comment by kraluk** [comment]
Ah, sorry for the problem... I did not pay any attention to that...

File: src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java [hunk 16]
```diff
@@ -1,41 +1,98 @@
...
+    @SneakyThrows
+    private void prepareXmlResponse(HttpServletResponse response, SyndFeed feed,
+                                    PrintWriter writer) {
+        response.setContentType(APPLICATION_ATOM_XML_VALUE);
+        new SyndFeedOutput().output(feed, writer);
+    }
+
+    @SneakyThrows
+    private void prepareBadRequestResponse(HttpServletResponse response, String format) {
+        response.sendError(BAD_REQUEST.value(),
```
Comment (airborn) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: I think that this should be 406 Not Acceptable

File: src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java [hunk 17]
```diff
@@ -1,41 +1,98 @@
...
-        @RequestParam(required = false) Set<String> excludedAuthors) {
-        limit = firstNonNull(limit, defaultLimit);
-        response.setContentType(MediaType.APPLICATION_ATOM_XML_VALUE);
-        new SyndFeedOutput().output(
-            rssProducer.getRss(request.getRequestURL().toString(), limit, excludedAuthors),
-            writer
+                       PrintWriter writer,
+                       @RequestParam(required = false) Integer limit,
+                       @RequestParam(required = false) Set<String> excludedAuthors,
+                       @RequestParam(required = false, defaultValue = "xml") String format) {
```
Comment (airborn) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: Current approach requires non standard parameter. Standard approach would be to use HTTP Content Negotiation (Accept header). I think that could be easily achieved by:

1. `@RequestMapping(path = RSS_FEED_MAPPING, produces = {MediaType.APPLICATION_JSON_UTF8_VALUE, MediaType.APPLICATION_ATOM_XML_VALUE})`
2. Returning `SyndFeed` from this method
3. Creating two beans of type `HttpMessageConverter` that would be responsible for serializing XML and JSON responses 

That way you would reuse existing framework mechanism and conform to standard HTTP content negotiation mechanism

File: src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java [hunk 18]
```diff
@@ -52,24 +60,39 @@ public void getRss(HttpServletRequest request, HttpServletResponse response,
         SyndFeed feed =
             rssProducer.getRss(request.getRequestURL().toString(), checkedLimit, excludedAuthors);
 
-        if (SupportedFormat.XML.toString().equalsIgnoreCase(format)) {
-            response.setContentType(MediaType.APPLICATION_ATOM_XML_VALUE);
+        Match(format).of(
```
Comment (tdziurko) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: Pattern matching, yay! :+1:

File: src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java (hunk 17 above)
Comment (tdziurko) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: Re @airborn I think that in most cases it is easier for the client (especially browser/other rss reader) to specify desired format using url parameter, so I would keep it this way.

Comment (airborn) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: Sorry, I should have probably mention this before, but content negotiation within Spring handles both cases (an one additional):

https://spring.io/blog/2013/05/11/content-negotiation-using-spring-mvc
http://www.baeldung.com/spring-mvc-content-negotiation-json-xml

Comment (kraluk) [inline src/main/java/com/jvm_bloggers/core/rss/BlogPostsController.java]: Roger that, thanks for the clarification - I'm working on changes.

**Comment by airborn** [comment]
@kraluk do you have nay update on this? It's bloking me with another feature I'd like to implement

**Comment by kraluk** [comment]
@airborn this weekend I'm going to finalize this task - sorry for the delay...

**Comment by kraluk** [comment]
And some new commits arrived - I have introduced content negotiation as @airborn suggested and I've decided to use path extension (i.e. /pl/rss.xml) instead of a GET parameter. As for me basically, it looks better in this form :)

**Comment by airborn** [comment]
I have two minor comments regarding current implementation. 

But we need to do something with the failing test... I think that the functionality of the controller is covered by the second test so in my opinion we can safely remove this one. @tdziurko what is your opinion on this?

File: src/main/java/com/jvm_bloggers/JvmBloggersWebConfiguration.java [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+public class JvmBloggersWebConfiguration extends WebMvcConfigurerAdapter {
+
+    @Override
+    public void configureContentNegotiation(final ContentNegotiationConfigurer configurer) {
+        configurer
+            .favorPathExtension(true)
+            .favorParameter(false)
+            .ignoreAcceptHeader(false)
+            .useJaf(true)
+            .defaultContentType(APPLICATION_JSON_UTF8)
```
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/JvmBloggersWebConfiguration.java]: We definitely should have done this. I'm going to commit fixes this afternoon.

File: src/main/java/com/jvm_bloggers/JvmBloggersWebConfiguration.java [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+@Configuration
+public class JvmBloggersWebConfiguration extends WebMvcConfigurerAdapter {
+
+    @Override
+    public void configureContentNegotiation(final ContentNegotiationConfigurer configurer) {
+        configurer
+            .favorPathExtension(true)
+            .favorParameter(false)
+            .ignoreAcceptHeader(false)
+            .useJaf(true)
```
Comment (kraluk) [inline src/main/java/com/jvm_bloggers/JvmBloggersWebConfiguration.java]: Yep, you're right - definitely my oversight.

**Comment by kraluk** [comment]
Fixes are now committed. As for the awkward test, I've removed it - if there will need to restore it, for sure I'll do this.

File: src/main/java/com/jvm_bloggers/core/rss/converters/SyndFeedXmlMessageConverter.java [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+            "Reading SyndFeed from a XML file is currently not supported!");
+    }
+
+    @Override
+    @SneakyThrows
+    protected void writeInternal(SyndFeed syndFeed, HttpOutputMessage outputMessage)
+        throws IOException, HttpMessageNotWritableException {
+
+        String xml = new SyndFeedOutput().outputString(syndFeed);
+        outputMessage.getBody().write(xml.getBytes());
```
Comment (airborn) [inline src/main/java/com/jvm_bloggers/core/rss/converters/SyndFeedXmlMessageConverter.java]: Sorry, I haven't noticed this before, but we can save some memory here. 
You can just create a `Writer` from `outputMessage` and pass it to `SyndFeedOutput`, something like:
```
Writer writer = new PrintWriter(outputMessage.getBody());
new SyndFeedOutpur().output());
```
This way we can skip conversion from Stream to String and back to Stream.

**Comment by kraluk** [comment]
@airborn, thanks for the hint! Changes are now committed.

**Pull request merged**